### PR TITLE
Benchmark performance

### LINKED ISSUE #682: Benchmark performance

### PR DESCRIPTION
**Pull request by cptartur**
Closes #682

## Introduced changes

- Added benchmarks for collecting, compiling and running tests using Critierion.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [ ] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/src/lib.rs [hunk 1]
```diff
@@ -41,11 +41,11 @@ use crate::collecting::{
 use crate::test_filter::TestsFilter;
 use test_collector::{FuzzerConfig, LinkedLibrary, RawForkConfig, RawForkParams, TestCase};
 
+pub mod collecting;
```
Comment (Arcticae) [inline crates/forge/src/lib.rs:44]: unrelated

File: crates/forge/benches/my_benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    assert_passed!(result);
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    let declare_and_interact_input = setup_declare_and_interact();
+    let collect_tests_input = setup_collect_tests();
+    let compile_tests_input = setup_compile_tests();
+
+    let mut group = c.benchmark_group("benchmark-normal-flow");
+    group.sampling_mode(SamplingMode::Flat); // https://bheisler.github.io/criterion.rs/book/user_guide/advanced_configuration.html#sampling-mode
```
Comment (Arcticae) [inline crates/forge/benches/my_benchmark.rs:186]:
```suggestion
    group.sampling_mode(SamplingMode::Flat); // (Relatively) Long running benchmark
```
Plz don't make people open up browser to find out what the code does

File: crates/forge/benches/my_benchmark.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        #[starknet::interface]
+        trait IHelloStarknet<TContractState> {
+            fn increase_balance(ref self: TContractState, amount: felt252);
+            fn decrease_balance(ref self: TContractState, amount: felt252);
+            fn get_balance(self: @TContractState) -> felt252;
+            fn interact_with_state(self: @TContractState) -> (felt252, felt252, felt252);
+        }
+
+        #[test]
+        fn declare_and_interact() {
```
Comment (Arcticae) [inline crates/forge/benches/my_benchmark.rs:103]:
```suggestion
        fn declare_deploy_and_interact() {
```

File: crates/forge/benches/my_benchmark.rs
Comment (Arcticae) [inline crates/forge/benches/my_benchmark.rs:1]: Please rename this file

File: crates/forge/benches/my_benchmark.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::time::Duration;
+use test_collector::LinkedLibrary;
+use test_utils::corelib::corelib_path;
+use test_utils::runner::{Contract, TestCase};
+use test_utils::running_tests::run_test_case;
+use test_utils::{assert_passed, test_case};
+
+fn setup_collect_tests() -> TempDir {
```
Comment (Arcticae) [inline crates/forge/benches/my_benchmark.rs:17]: Can we split the cases into different files? So that extending/reading this suite is easier?

**Review by Arcticae — COMMENTED** [review]
No major objections.
When we will perform those benchmarks? 
I kind of am lacking overall usage scenario/normalized flow and i'm afraid that this will quickly become obsolete because of this

**Comment by cptartur** [comment]
> No major objections. When we will perform those benchmarks? I kind of am lacking overall usage scenario/normalized flow 

My intention was for this to mostly be used locally by devs: This allows verifying if changes being introduced to benchmarked parts of forge do not cause performance regression.

We could introduce these as some part of release flow, CI, or similar, though I'm unsure if that's necessary.

> and i'm afraid that this will quickly become obsolete because of this

Fair point: In terms of pure syntax problems, lint checks should enforce these files being up to date. Cairo code in `declare_and_interact` is more problematic. Perhaps we should move it to a separate file and also run a standard test agains it so it stays up to date? 

In terms of keeping this "relevant", not sure how we can tackle this. In case devs don't use these benchmarks they may become obsolete.

**Review by piotmag769 — APPROVED** [review]
I wonder if we should also add a point to our checklist in our PR template sth like 
`- [ ] Ran benchmarks (if changes can affect performance)`
so it doesn't become obsolete

File: Cargo.toml [hunk 5]
```diff
@@ -58,6 +58,7 @@ tempfile = "3.8.0"
 thiserror = "1.0.49"
 ctor = "0.2.4"
 url = "2.4.0"
+
```
[the remark this answers is not in the document]
Comment (cptartur) [reply Cargo.toml:61]: Added by accident

**Comment by cptartur** [comment]
> I wonder if we should also add a point to our checklist in our PR template sth like `- [ ] Ran benchmarks (if changes can affect performance)` so it doesn't become obsolete

We could do that, but I'm unsure if that will solve the issue at all.

**Comment by MaksymilianDemitraszek** [comment]
> Perhaps we should move it to a separate file and also run a standard test agains it so it stays up to date
IMO this is strongly required (can be done in a separate PR.

**Comment by cptartur** [comment]
> > Perhaps we should move it to a separate file and also run a standard test agains it so it stays up to date
> 
> IMO this is strongly required (can be done in a separate PR)

I've included it in this PR already

**Comment by piotmag769** [comment]
I think it's ready to merge and we can update the PR template after? @cptartur

**Comment by cptartur** [comment]
> I think it's ready to merge and we can update the PR template after? @cptartur

#1044
